python/mpt/src/mpt/trie_diff.py

```diff
@@ -0,0 +1,201 @@
+        r_root = tries.post_state_root
+
+        diff._compute_diff(l_root, r_root, Bytes())
+        return diff
+
+    def _compute_diff(self, left: Optional[Hash32], right: Optional[Hash32], path: Bytes):
+        if left == right:
+            return
+
+        l_node = self._nodes.get(left) if left else None
+        r_node = self._nodes.get(right) if right else None
+
+        l_name = l_node.__class__.__name__ if l_node else "None"
+        r_name = r_node.__class__.__name__ if r_node else "None"
+
+        match l_name, r_name:
+            case "None", "None":
+                # No change
+                pass
+
+            case "None", "LeafNode":
+                # new leaf
+                preimage = nibble_path_to_bytes(path + r_node.rest_of_key)
+                address = self._address_preimages[preimage]
+                self._main_trie[address] = tuple((None, AccountNode.from_rlp(r_node.value)))
+
+            case "None", "ExtensionNode":
+                r_node: ExtensionNode = cast(ExtensionNode, r_node)
+                # Look for diffs in the right sub-tree
+                self._compute_diff(None, r_node.subnode, path + r_node.key_segment)
+
+            case "None", "BranchNode":
+                # Look for diffs in all branches of the right sub-tree
+                r_node: BranchNode = cast(BranchNode, r_node)
+                for i in range(0, 16):
+                    self._compute_diff(None, r_node.subnodes[i], path + i.to_bytes(1, "big"))
+
+            case "LeafNode", "None":
+                # deleted leaf (should not happen post-cancun)
+                preimage = nibble_path_to_bytes(path + l_node.rest_of_key)
+                address = self._address_preimages[preimage]
+                self._main_trie[address] = tuple((AccountNode.from_rlp(l_node.value), None))
+
+            case "LeafNode", "LeafNode":
+                if l_node.value != r_node.value:
+                    preimage = nibble_path_to_bytes(path + l_node.rest_of_key)
+                    address = self._address_preimages[preimage]
+                    self._main_trie[address] = tuple((AccountNode.from_rlp(l_node.value), AccountNode.from_rlp(r_node.value)))
+
+            case "LeafNode", "BranchNode":
+                # Left is a valute, right is
+                logger.warning("LeafNode -> BranchNode")
+                breakpoint()
+                for i in range(0, 16):
+                    self._compute_diff(None, r_node.subnodes[i], path + i.to_bytes(1, "big"))
+
+            case "LeafNode", "ExtensionNode":
+                raise ValueError("LeafNode -> ExtensionNode should not happen if comparing two valid trie roots")
+
+            case "ExtensionNode", "None":
+                # Look for diffs in the left sub-tree
+                l_node: ExtensionNode = cast(ExtensionNode, l_node)
+                self._compute_diff(l_node.subnode, None, path + l_node.key_segment)
+
+            case "ExtensionNode", "LeafNode":
```

Comment:
left: ext -> branch -> leaf A & leaf B
right: deletion of leaf A, then branch converted into ext, ext to leaf converted into unique leaf